dmoj/judge.py

```diff
@@ -284,22 +226,235 @@ def listen(self):
+            if exc:
+                try:
+                    raise exc
+                except KeyboardInterrupt:
+                    # Let KeyboardInterrupt bubble up.
+                    raise
+                except:  # noqa: E722, we want to catch everything
+                    pass
+
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+
+        # Strip ANSI from the message, since this might be a checker's CompileError
+        # ...we don't want to see the raw ANSI codes from GCC/Clang on the site.
+        # We could use format_ansi and send HTML to the site, but the site doesn't presently support HTML
+        # internal error formatting.
+        self.packet_manager.internal_error_packet(strip_ansi(message))
+
+        # Logs can contain ANSI, and it'll display fine
+        print(message, file=sys.stderr)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        pass
+
+
+class SubmissionJudgeProcess:
+    def __init__(self, submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta):
+        self.submission_id = submission_id
+        self.problem_id = problem_id
+        self.language = language
+        self.source = source
+        self.time_limit = time_limit
+        self.memory_limit = memory_limit
+        self.short_circuit = short_circuit
+        self.meta = meta
+
+        self._abort_requested = False
+        self.grader = None
+        self.process = None
+        self.process_conn = None
+
+    def run(self):
+        self.process_conn, child_conn = multiprocessing.Pipe()
+        self.process = multiprocessing.Process(
+            name="DMOJ Judge Handler for %s/%d" % (self.problem_id, self.submission_id),
+            target=self._run_grading_wrapper_process,
+            args=(child_conn,),
+        )
+        self.process.start()
+        child_conn.close()
+
+        while True:
+            try:
+                ipc_msg = self.process_conn.recv()
+            except Exception:
+                logger.error("Failed to read IPC message from handler!")
+                raise
+
+            if ipc_msg is None:
+                self.process_conn.send(None)
+                break
+
+            yield ipc_msg
+
+        self.process_conn.close()
+        self.process.join()
+
+    def stop(self):
+        if self.process:
+            # Might be None if run was never called, or failed.
+            try:
+                self.process.join(timeout=IPC_TEARDOWN_TIMEOUT)
+                if self.process.is_alive():
+                    self.process.kill()
+                    logger.error("Judge IPC process is still alive after timeout, sending SIGKILL!")
+            except OSError:
+                logger.exception('Exception while waiting for judge IPC process to shut down, ignoring...')
+
+    def abort_grading(self):
```

Comment:
What if an abortion was received _before_ `self.process` is assigned?

Reply:
I don't even know why this is guarded by `self.process`... it doesn't have to be.

I've added an `ipc_ready_signal` event though, so that `begin_grading` doesn't return before the IPC process is created, so we remove the race where an abort arrives while the IPC process is spawning.